### PR TITLE
Readme examples

### LINKED ISSUE #33: Add complete examples in README/wiki

### PR DESCRIPTION
**Pull request by apostergiou**
- Introduce `_examples` directory
This directory can be used to hold the source code for usage examples.
For starters, it will contain a bundler project which builds Ruby gems.

- README: Add complete, runnable examples

Closes #33

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +54,109 @@ $ go get -u github.com/skroutz/mistry/cmd/mistry
 
 
 
+Examples
```
Comment (agis) [inline README.md]: We could move this whole section under a README in the `examples` directory and just add a link to that readme in here:

> For example projects take a look at the [examples](examples/) directory.

File: README.md [hunk 2]
```diff
@@ -54,6 +54,109 @@ $ go get -u github.com/skroutz/mistry/cmd/mistry
...
+--------------------------------------------------
+Here is a little preview of how building looks with mistry.
+
+### Bundler
```
Comment (agis) [inline README.md]: This could be a README under each project's directory.

Also, I believe we should rename it to `bundler-cache` so that it's more apparent what this is about.

File: README.md [hunk 3]
```diff
@@ -54,6 +54,109 @@ $ go get -u github.com/skroutz/mistry/cmd/mistry
...
+
+```sh
+$ mkdir -p /tmp/mistry/projects
+```
+
+3. Copy the example projects to the projects directory
+
+```sh
+$ cp -r _examples/* /tmp/mistry/projects/
+```
```
Comment (agis) [inline README.md]: Since these are a one-time setup instructions, they should probably be in a wiki page (ie. "Setup") and just in `examples/README.md` that having completed the "Setup" step is a prerequisite.

File: _examples/bundler/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    cmake \
+    libcurl4-openssl-dev \
+    libicu-dev \
+    libmagic-dev \
+    libmariadbclient-dev \
+    libsqlite3-dev \
+    libsystemd-dev \
+    pkg-config \
+    ruby ruby-dev ruby-bundler \
+    zlib1g-dev
```
Comment (agis) [inline _examples/bundler/Dockerfile]: Are we sure all these dependencies are needed? icu, libmagic, libmariadbclient etc. seem irrelevant.

File: _examples/bundler/docker-entrypoint [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env ruby
+require 'open3'
+
+# non-buffered stdout
+STDOUT.sync = true
+
+GEMFILE = File.read("/data/params/Gemfile").strip
```
Comment (agis) [inline _examples/bundler/docker-entrypoint]: Since we're in data, the `/data/` part isn't needed.

File: _examples/bundler/docker-entrypoint [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+abort "Gemfile param cannot be empty" if GEMFILE.empty?
+
+GEMFILE_LOCK = File.read("/data/params/Gemfile.lock").strip
```
Comment (agis) [inline _examples/bundler/docker-entrypoint]: Since we're in data, the `/data/` part isn't needed.

File: _examples/bundler/docker-entrypoint [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+abort "Gemfile.lock param cannot be empty" if GEMFILE_LOCK.empty?
+
+cmd = "bundle install --path /data/artifacts --no-prune --deployment --jobs=4 --gemfile=/data/params/Gemfile"
```
Comment (agis) [inline _examples/bundler/docker-entrypoint]: Since we're in data, the `/data/` part isn't needed.

File: _examples/bundler/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+WORKDIR /data
+
+COPY docker-entrypoint /usr/local/bin/docker-entrypoint
+RUN chmod +x /usr/local/bin/docker-entrypoint
+
+ENTRYPOINT ["/usr/local/bin/docker-entrypoint"]
```
Comment (agis) [inline _examples/bundler/Dockerfile]: Any reason why it's `_examples/` and not `examples/`?

Comment (apostergiou) [reply _examples/bundler/Dockerfile]: The `_examples/` directory can act as a temporary directory, it is not necessary for building with mistry. The `go/build` package also ignores these files https://golang.org/pkg/go/build/#Context.Import.

File: _examples/bundler/docker-entrypoint (hunk 7 above)
Comment (apostergiou) [reply _examples/bundler/docker-entrypoint]: This is needed, otherwise bundler uses the `--gemfile` option path as the root and create a tree like this:

```sh
├── artifacts
├── cache
└── params
    ├── artifacts
    │   └── ruby
    │       └── 2.3.0
    ├── Gemfile
    └── Gemfile.lock
```

File: _examples/bundler/docker-entrypoint (hunk 6 above)
Comment (apostergiou) [reply _examples/bundler/docker-entrypoint]: fixed

File: _examples/bundler/docker-entrypoint (hunk 5 above)
Comment (apostergiou) [reply _examples/bundler/docker-entrypoint]: fixed

File: _examples/bundler/Dockerfile (hunk 4 above)
Comment (apostergiou) [reply _examples/bundler/Dockerfile]: Some of them are redundant indeed.

File: README.md (hunk 3 above)
Comment (apostergiou) [reply README.md]: nice, updated

File: README.md (hunk 2 above)
Comment (apostergiou) [reply README.md]: sounds good, updated

File: README.md (hunk 1 above)
Comment (apostergiou) [reply README.md]: done
